mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,566 @@
+                continue
+            normalized_text = self._normalize_text(raw_token["text"])
+            # We tokenize the post-norm text and split the entity if possible
+            # Ex: normalize("o'clock") -> "o clock" -> ["o", "clock"]
+            normalized_texts = [t["text"] for t in self.tokenize(normalized_text)]
+
+            if len(normalized_texts) > 0:
+                for token_text in normalized_texts:
+                    normalized_tokens.append(
+                        {
+                            "entity": token_text,
+                            "raw_entity": raw_token["text"],
+                            "raw_token_index": i,
+                            "raw_start": raw_token["start"],
+                        }
+                    )
+        return normalized_tokens
+
+    def get_normalized_tokens_as_tuples(self, text):
+        """Gets normalized tokens from input text and returns the result as a tuple.
+
+        Args:
+            text (str): Text to normalize.
+        Returns:
+            normalized_tokens_as_tuples (Tuple(str)): A Tuple of normalized tokens.
+        """
+        return tuple(t["entity"] for t in self.tokenize_and_normalize(text))
+
+    def stem_word(self, word):
+        """
+        Gets the stem of a word. For example, the stem of the word 'fishing' is 'fish'.
+
+        Args:
+            words (List[str]): List of words to stem.
+
+        Returns:
+            stemmed_words (List[str]): List of stemmed words.
+        """
+        return self.stemmer.stem_word(word)
+
+    @staticmethod
+    def find_mindmeld_annotation_re_matches(text):
+        """
+        Args:
+            text (str): The string to find mindmeld annotation instances
+                (" {entity_text|entity_type} ")
+        Returns:
+            matches (List[sre.SRE_Match object]): Regex match objects.
+        """
+        return list(MINDMELD_ANNOTATION_PATTERN.finditer(text))
+
+    @staticmethod
+    def modify_around_annotations(text, function, keep_space_before_annotation=False):
+        """Applied a function around the mindmeld annotation.
+
+        function(pre_entity_text) + { + function(entity_text) + |entity_name}
+            + function(post_entity_text)
+
+        Args:
+            text (str): Original sentence with markup to modify.
+            function (function): Function to apply around the annotation
+            keep_space_before_annotation (bool): If True and if there is a space before entity
+                annotation, then the space will not be modified.
+        Returns:
+            modified_text (str): Text modified around annotations.
+        """
+        matches = TextPreparationPipeline.find_mindmeld_annotation_re_matches(text)
+
+        modified_text = []
+        prev_entity_end = 0
+
+        for match in matches:
+            entity_start, entity_end = match.span()
+
+            # Include Space Before Entity Annotation
+            if (
```

Comment:
What happens when there are multiple spaces like "xyz      {abc}". I think in this case, we need to preserve all of the spaces.

Reply:
What is the motivation for preserving multiple spaces? From my understanding, a single space was preserved so that MindMeld entity annotations can be detected downstream in the application. If that is the reason, we don't necessarily need multiple spaces?

Reply:
Preprocessors now process the input text directly